agent/engine/docker_task_engine.go

```diff
@@ -245,53 +245,73 @@ func (engine *DockerTaskEngine) synchronizeState() {
-				// Figure out the dockerid
-				describedCont, err := engine.client.InspectContainer(cont.DockerName, inspectContainerTimeout)
-				if err != nil {
-					log.Warn("Could not find matching container for expected", "name", cont.DockerName)
-				} else {
-					cont.DockerID = describedCont.ID
-					cont.Container.SetKnownStatus(dockerStateToState(describedCont.State))
-					// update mappings that need dockerid
-					engine.state.AddContainer(cont, task)
-					engine.imageManager.RecordContainerReference(cont.Container)
-				}
-			}
-			if cont.DockerID != "" {
-				currentState, metadata := engine.client.DescribeContainer(cont.DockerID)
-				if metadata.Error != nil {
-					currentState = api.ContainerStopped
-					if !cont.Container.KnownTerminal() {
-						cont.Container.ApplyingError = api.NewNamedError(&ContainerVanishedError{})
-						log.Warn("Could not describe previously known container; assuming dead", "err", metadata.Error, "id", cont.DockerID, "name", cont.DockerName)
-						engine.imageManager.RemoveContainerReferenceFromImageState(cont.Container)
-					}
-				} else {
-					engine.imageManager.RecordContainerReference(cont.Container)
-					if engine.cfg.ContainerMetadataEnabled && !cont.Container.IsMetadataFileUpdated() {
-						go engine.updateMetadataFile(task, cont)
-					}
-
-				}
-				if currentState > cont.Container.GetKnownStatus() {
-					cont.Container.SetKnownStatus(currentState)
-				}
-			}
+			engine.synchronizeContainerStatus(cont, task)
+		}
+
+		tasksToStart = append(tasksToStart, task)
+
+		// Put tasks that are stopped by acs but hasn't been stopped in wait group
+		if task.GetDesiredStatus().Terminal() && task.GetStopSequenceNumber() != 0 {
+			engine.taskStopGroup.Add(task.GetStopSequenceNumber(), 1)
 		}
+	}
+
+	for _, task := range tasksToStart {
 		engine.startTask(task)
 	}
+
 	engine.saver.Save()
 }
 
+// synchronizeContainerStatus checks and updates the container status with docker
+func (engine *DockerTaskEngine) synchronizeContainerStatus(container *api.DockerContainer, task *api.Task) {
+	if container.DockerID == "" {
+		log.Debug("Found container potentially created while we were down", "name", container.DockerName)
+		// Figure out the dockerid
+		describedContainer, err := engine.client.InspectContainer(container.DockerName, inspectContainerTimeout)
+		if err != nil {
+			log.Warn("Could not find matching container for expected", "name", container.DockerName)
+		} else {
+			container.DockerID = describedContainer.ID
+			container.Container.SetKnownStatus(dockerStateToState(describedContainer.State))
+			// update mappings that need dockerid
+			engine.state.AddContainer(container, task)
+			engine.imageManager.RecordContainerReference(container.Container)
+		}
+	}
+
+	if container.DockerID != "" {
+		currentState, metadata := engine.client.DescribeContainer(container.DockerID)
+		if metadata.Error != nil {
+			currentState = api.ContainerStopped
```

Comment:
This is a bug in the Agent today, where if we can't inspect a container we mark it as stopped. Please add a `TODO` here to come back and fix this (better yet, create an issue). @samuelkarp would have more details about the bug.

Reply:
I don't think we want to assume that it's dead.  However, I do think it would make more sense to mark it as desired=STOPPED and then let it progress through the task engine.  I'm looking at a bug right now where we do something very similar and assume that a container is stopped when it isn't, which can cause problems (resource conflicts, unintended processes still running, etc).